File: 2022/day6/day6.py
from typing import List, Optional


def getInput(useTest) -> List[str]:
    inputFile = "input-test.txt" if useTest else "input.txt"
    with open(inputFile, "r") as file1:
        return [x.strip() for x in file1]


def main(useTest: Optional[bool] = False) -> None:
    inputData = getInput(useTest)

    for line in inputData:
        print("day 6 part 1:", getStartofPacket(line))
        print("day 6 part 2:", getStartofPacket(line, 14))


def getStartofPacket(word: str, length: int = 4) -> int:
    letters = []
    for i, char in enumerate(word):
        letters.append(char)
        if len(letters) >= length:
            if isUnique(letters) and len(letters):
                return i + 1
            else:
                letters.pop(0)
    return 0


def isUnique(word: List[str]) -> bool:
    letters = [*range(1, 26)]
    for char in word:
        charNum = ord(char) - ord("a")
        if charNum in letters:
            letters.pop(letters.index(charNum))
        else:
            return False
    return True


if __name__ == "__main__":
    main()

File: 2021/day6/day6.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Hydrothermal struct {
	useTest   bool
	fishes    []int
	birth     int
	totalDays int
}

type point struct {
	x, y int
}

func (this *Hydrothermal) getInput() {
	inputFile := "input.txt"
	if this.useTest {
		inputFile = "input-test.txt"
	}

	file, _ := os.Open(inputFile)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		splitted := strings.Split(line, ",")
		for _, s := range splitted {
			n, _ := strconv.Atoi(s)
			this.fishes = append(this.fishes, n)
		}

	}
	this.birth = 7
	this.totalDays = 256
	if this.useTest {
		this.totalDays = 256
	}
	defer file.Close()
}

func (this *Hydrothermal) countSpawn(days int, remain int) int {
	if days < 7 || days < remain {
		return 0
	}
	days -= remain
	count := 1
	newFish := days / this.birth
	for i := 0; i <= newFish; i++ {
		count += this.countSpawn(days, 9)
		days -= this.birth

	}
	return count + newFish
}

func (this *Hydrothermal) countFish() int {
	var count int
	cache := make(map[int]int)
	for i, fish := range this.fishes {
		fmt.Println("here", i)
		count += this.countSpawn(this.totalDays, fish)
	}

	return count + len(this.fishes)
}

func main() {
	Hydrothermal := &Hydrothermal{
		useTest: false,
	}
	Hydrothermal.getInput()
	fmt.Println("Day 5 part 1:", Hydrothermal.countFish())
	fmt.Println("Day 5 part 2:", Hydrothermal.countFish())
}
